state/protocol/badger/mutator.go

```diff
@@ -556,17 +556,39 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
 
 	// track service event driven metrics and protocol events that should be emitted
 	var events []func()
+
+	// track service events to index results by event type and height
+	var serviceEvents []struct {
+		height   uint64
+		resultID flow.Identifier
+		event    flow.ServiceEvent
+	}
+
 	for _, seal := range parent.Payload.Seals {
-		// skip updating epoch-related metrics if EECC is triggered
-		if epochFallbackTriggered {
-			break
-		}
 
 		result, err := m.results.ByID(seal.ResultID)
 		if err != nil {
 			return fmt.Errorf("could not retrieve result (id=%x) for seal (id=%x): %w", seal.ResultID, seal.ID(), err)
 		}
 		for _, event := range result.ServiceEvents {
+			// it is possible, however very unlikely, that there are multiple service events of the same type in a single result.
+			// This will add multiple identical entries here, but the storage layer will only keep one index entry per combination of
+			// height and event type, essentially disregarding duplicates. Since the situation is very unlikely and other layer will handle
+			// this situation gracefully we leave it like this instead of writing deduplication logic here.
+			serviceEvents = append(serviceEvents, struct {
+				height   uint64
+				resultID flow.Identifier
+				event    flow.ServiceEvent
+			}{
+				height:   parent.Header.Height,
```

Comment:
It's still not clear to me why the result is indexed by the height of the parent block here?

Shouldn't we index a result for a height where it comes to effect?

We'd better add some comments or examples to explain what height is being used, and why.

For instance, if a service event, which indicates VersionUpdate from v1 to v2 at height 2000, is included in a result included in block 100. Do we index the VersionUpdate event by height 100 or 2000?

Reply:
That's to be consistent with how service event effectiveness is described in https://github.com/onflow/flow-go/blob/master/state/protocol/badger/mutator.go#L759 - so after a QC is there for finalized block.

This PR shouldn't contain any mentions of VersionUpdate events, but general purpose and reasoning has been explained https://github.com/onflow/flow-go/pull/3605/files#diff-c760f8136a91a3d5ffbcfa069cf895ce0e65f4a60b1df9ad8cda5dbd04211c55R29

Reply:
I think the index height here should be `header.Height` - the height of the block being inserted. 

If a service event is coming into effect at the block we are processing, then that block has the form of block D (definition linked below). The parent of D (C) is the block containing the relevant seal, so while processing D we iterate over C's seals. But the point at which the service event comes into effect is still block D (not C), so we should use `header.Height` (D's height) rather than `parent.Height` (C's height).

https://github.com/onflow/flow-go/blob/b8561103450733edaa270131932dc555e8513360/state/protocol/badger/mutator.go#L749-L757

Reply:
Fixed